src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -62,20 +62,41 @@ internal class PackageRestoreDataSource : ChainedProjectValueDataSourceBase<Rest
         private readonly IFileSystem _fileSystem;
         private readonly IProjectDiagnosticOutputService _logger;
         private readonly IProjectDependentFileChangeNotificationService _projectDependentFileChangeNotificationService;
-
+        private readonly IVsSolutionRestoreService4 _solutionRestoreService4;
         private byte[]? _latestHash;
         private bool _enabled;
 
+        /// <summary>
+        /// Project Unique Name used by Nuget Nomination.
+        /// </summary>
+        public string Name => _project.FullPath;
+
+        /// <summary>
+        /// Re-usable task that completes when there is a new nomination
+        /// </summary>
+        private TaskCompletionSource<bool>? _whenNominatedTask;
+
+        private bool _nugetRestoreStarted;
+
+        private bool _wasSourceBlockContinuationSet;
+
+        /// <summary>
+        /// Save the configured project versions that might get nominations.
+        /// </summary>
+        private readonly Dictionary<ProjectConfiguration, IComparable> _savedNominatedConfiguredVersion = new();
+
         [ImportingConstructor]
         public PackageRestoreDataSource(
             UnconfiguredProject project,
             IPackageRestoreUnconfiguredInputDataSource dataSource,
-            [Import(ExportContractNames.Scopes.UnconfiguredProject)]IProjectAsynchronousTasksService projectAsynchronousTasksService,
+            [Import(ExportContractNames.Scopes.UnconfiguredProject)] IProjectAsynchronousTasksService projectAsynchronousTasksService,
             IVsSolutionRestoreService3 solutionRestoreService,
             IFileSystem fileSystem,
             IProjectDiagnosticOutputService logger,
-            IProjectDependentFileChangeNotificationService projectDependentFileChangeNotificationService)
-            : base(project, synchronousDisposal : true, registerDataSource : false)
+            IProjectDependentFileChangeNotificationService projectDependentFileChangeNotificationService,
+            IVsSolutionRestoreService4 solutionRestoreService4,
```

Comment:
Does IVsSolutionRestoreService4 actually a different instance of IVsSolutionRestoreService3? Do we need two fields to represent the same underlying service?

Reply:
IVsSolutionRestoreService4 does not inherit from IVsSolutionRestoreService3 or others.